### DISCUSSION TITLE
vue-tsc support for js emit

### DISCUSSION THREAD

**Discussion by DarrenTheDev** [vuejs/language-tools#640] [Q&A] [answered]
I'm looking to distribute a component library built with Vue3 / TS in single file components. I'd like for the consuming application to have access to the types when using the components. An example of a library that does this well is [Naive UI](https://github.com/TuSimple/naive-ui). It has components written in .tsx files which built into individual .js and .d.ts files with the `tsc -b` command.

Attempting to do the above with .vue files by running `vue-tsc -b` throws the message that "js emit is not support". Is this a feature that might be supported in the future to make this use case possible?

**Comment by ghost** [comment, marked as answer]
Not familiar with this case but maybe you could just generate .d.ts files with vue-tsc, then compile .vue to .js files with Vite?

Even with Vue 2 I used to build libs like (vue cli): `vue-cli-service build --target lib && tsc`

**Reply by ghost** [reply]
To elaborate — "emit JS" doesn't really make sense for .vue files, because they're not "just javascript" — contrary to `.jsx`

TSX + **emit** JS = JSX + D.TS files
.vue lang="ts" + **build** + **run vue-tsc** = JS + D.TS files
.vue lang="ts" + **emit** JS = not applicable

.vue files must be compiled to javascript, whereas .tsx files can be just stripped of the TypeScript specific stuff to become JS which I assume is the main way "emitJs" works.

To extend my first post, here's how to do the "build + run vue-tsc" if you want to publish your own package

Here's a working example that uses vue-tsc AND outputs advanced typing information:
![image](https://user-images.githubusercontent.com/5359825/139333606-702604b4-9a37-4a2f-ab43-a205f68e245b.png)
![image](https://user-images.githubusercontent.com/5359825/139333539-2d4f441b-4f72-411f-b142-e93f9260af4f.png)
![image](https://user-images.githubusercontent.com/5359825/139333561-12723670-8ab3-457b-b86b-902723aedca0.png)


[repo](https://github.com/sethidden/vue-tsc-component-library-example)
[npm package](https://www.npmjs.com/package/@3nuc/vue-tsc-component-library-example)

There are a few caveats:
* You need to remove the TS shim that makes all .vue files a dumb TS type. Just remove the `./src/env.d.ts` from your component library, or else all the detailed vue-tsc typings will be thrown away (env.d.ts will take precedence)
 
* Set the `build` script in your component library's `package.json` to: `vue-tsc --declaration --emitDeclarationOnly && vite build` (see [vue-tsc readme](https://github.com/johnsoncodehk/volar/blob/master/packages/vue-tsc/README.md?plain=1#L24)). Make sure vue-tsc is a recent version (eg. [`npm init vue@next`](https://github.com/vuejs/create-vue) has a very old version so you need to update or you'll get a `not support` error when you run the above)

If you have any questions feel free to ask.

**Reply by DarrenTheDev** [reply]
Thanks so much for your detailed help here! I had been down that route, but the missing piece of the puzzle was the `typings` attribute in the `package.json` file to point typescript to the definition for the entry file.

Here's what I ended up with in the package.json for anyone coming across this in the future. I set the outDir in the vite.config.ts to dist/lib and in the tsconfig.json to dist/types.

```json
{
  "name": "library",
  "version": "0.0.0",
  "main": "./dist/lib/library.umd.js",
  "module": "./dist/lib/library.es.js",
  "typings": "./dist/types/main.d.ts",
  ...
}
```

**Reply by Jackie1210** [reply]
https://github.com/johnsoncodehk/volar/issues/1496 @sethidden  have you ever meet this issue like this?

**Reply by reslear** [reply]
ok, vue-tsc generate types, vite clean folder :) 
need only swap, first vite clean folder and generate files, then vue-tsc generate types

```sh
vite build && vue-tsc --declaration --emitDeclarationOnly
```

:)

**Reply by andredewaard** [reply]
@DarrenTheDev Thanks! So its not possible to just compile the individual vue files into js files and consume those? 
I want something like [this](https://cmdcolin.github.io/posts/2022-05-27-youmaynotneedabundler)  but with vue-tsc

**Reply by codethief** [reply]
@andredewaard Instead of writing .vue files you *could* define your Vue components in regular TSX files, see https://vuejs.org/guide/extras/render-function.html#jsx-tsx and, in particular, the [code example at the very bottom of the page](https://vuejs.org/guide/extras/render-function.html#typing-functional-components). (TL;DR This makes Vue components look quite similar to React components.)

**Reply by heyakyra** [reply]
[sethidden](https://github.com/sethidden) on Oct 28, 2021:
> [repo](https://github.com/sethidden/vue-tsc-component-library-example) [npm package](https://www.npmjs.com/package/@3nuc/vue-tsc-component-library-example)

This dead link is referred to by the [`README`](https://github.com/vuejs/language-tools/tree/master/packages/tsc#vue-tsc) so I tried to update it ~but it won't allow me to open a pull request: https://github.com/vuejs/language-tools/compare/heyakyra/master~ 

**Edit:** Nevermind, got a PR open! https://github.com/vuejs/language-tools/pull/5721

cc: @johnsoncodehk @KazariEX

**Comment by messenjer** [comment]
Hello,

This morning I was trying to do the same thing than @DarrenTheDev (to distribute a component library built with Vue3 / TS in single file components.)

I found your discussion, and I tested the final solution from @DarrenTheDev where we replace also the:

* vite.config.js
* tsconfig.json
* package.json

To have at the end, all the *.d.ts files in the dist folder.

It worked, but at the end I found another solution, the [vite-plugin-dts](https://github.com/qmhc/vite-plugin-dts)  

@DarrenTheDev have you tested it?

**Reply by DarrenTheDev** [reply]
Thanks Mathieu I hadn't come across that package. 

Looks to work pretty well in my testing and gives pretty well the same result.

**Reply by johnsoncodehk** [reply]
Normally, vue-tsc and other solutions get similar results. The only difference is that vue-tsc reuses the logic of volar processing script setup, so vue-tsc can handle such as `defineProps<{ ... }>()` , `defineEmit<{ ... }>()` syntax and resolves to the expected component type.

**Reply by ghost** [reply]
@johnsoncodehk Do you know what does vite-plugin-dts use under the hood to scan the .vue files and emit their typedefs? I thought vue-tsc was the only way

**Comment by messenjer** [comment]
Thank you very much for the answer.

I see that there is also the support for the script setup, `defineProps<{...}>()` in vite-plugin-dts https://github.com/qmhc/vite-plugin-dts/commit/1b62755947972a67a034830cc7478fa1a65924c4 

At the moment, I don't use `script setup` anymore but `defineComponents` because Storybook (with vue-docgen-api) `defineProps` is not well supported. https://github.com/vue-styleguidist/vue-styleguidist/issues/997  

I would love to switch back to `script setup`.

**Comment by ccqgithub** [comment]
That seems all not work with `slots` (with `volar`)?

I can get the `sot props` type when I directly import the `.vue` file. 

<img width="582" alt="image" src="https://user-images.githubusercontent.com/1654761/162567406-c53ed7c2-ff03-4334-b21b-43a090e392af.png">

But if I build a lib, then import the component form the lib, the slot props type is `any`.

<img width="452" alt="image" src="https://user-images.githubusercontent.com/1654761/162567442-9828306a-7359-4c2a-bcb4-feb9a6284eb5.png">

The `slot props type` lost in the declaration files for component?

And, their type is different in the `volar`.

<img width="771" alt="image" src="https://user-images.githubusercontent.com/1654761/162567718-97c1c246-86ec-4a11-bc41-e38a400fc398.png">

From lib:
<img width="724" alt="image" src="https://user-images.githubusercontent.com/1654761/162567713-8283995e-4f5e-46bb-9a06-fc6f6f4ca59b.png">

**Reply by ghost** [reply]
This could indeed be a bug, though the underlying TS service should be the same for both vue-tsc and volar. Are you using the latest vue-tsc in package.json? `npm list vue-tsc`. It could be your vscode has updated volar automatically to the latest version (works) but your NPM vue-tsc stayed old or something.

**Reply by ccqgithub** [reply]
> This could indeed be a bug, though the underlying TS service should be the same for both vue-tsc and volar. Are you using the latest vue-tsc in package.json? `npm list vue-tsc`. It could be your vscode has updated volar automatically to the latest version (works) but your NPM vue-tsc stayed old or something.

I just install the latest version of `vue-tsc@0.34.2`, the problem is the same.

After I run `vue-tsc --declaration --emitDeclarationOnly`, the output of `.vue` like below, no slots info?

```ts
// Follow.vue.d.ts
declare const _default: import("vue").DefineComponent<{}, {}, {}, {}, {}, import("vue").ComponentOptionsMixin, import("vue").ComponentOptionsMixin, import("vue").EmitsOptions, string, import("vue").VNodeProps & import("vue").AllowedComponentProps & import("vue").ComponentCustomProps, Readonly<import("vue").ExtractPropTypes<{}>>, {}>;
export default _default;
//# sourceMappingURL=Follow.vue.d.ts.map
```

And if I use `vite-plugin-dts`, output like this:

```ts
// Follow.vue.d.ts
import { ComponentPublicInstance, StyleValue } from 'vue';
declare const _sfc_main: import("vue").DefineComponent<{}, {
    parent: ResizeObserverEntry | null;
    target: import("vue").Ref<ResizeObserverEntry | null>;
    setTargetEl: (elm: Element | HTMLElement | ComponentPublicInstance<{}, {}, {}, {}, {}, {}, {}, {}, false, import("vue").ComponentOptionsBase<any, any, any, any, any, any, any, any, any, {}>> | null) => void;
    popup: import("vue").Ref<ResizeObserverEntry | null>;
    setPopupEl: (elm: Element | HTMLElement | ComponentPublicInstance<{}, {}, {}, {}, {}, {}, {}, {}, false, import("vue").ComponentOptionsBase<any, any, any, any, any, any, any, any, any, {}>> | null) => void;
    style: import("vue").ComputedRef<StyleValue>;
    slots: Readonly<{
        [name: string]: import("vue").Slot | undefined;
    }>;
}, {}, {}, {}, import("vue").ComponentOptionsMixin, import("vue").ComponentOptionsMixin, import("vue").EmitsOptions, string, import("vue").VNodeProps & import("vue").AllowedComponentProps & import("vue").ComponentCustomProps, Readonly<import("vue").ExtractPropTypes<{}>>, {}>;
export default _sfc_main;
//# sourceMappingURL=Follow.vue.d.ts.map
```

And when I use the latest version of `vue-tsc`, has other .d.ts files with `__VLS_`

<img width="348" alt="image" src="https://user-images.githubusercontent.com/1654761/162611757-b701211c-2df8-4096-821c-57a7ba21e7c9.png">

**Reply by ccqgithub** [reply]
> This could indeed be a bug, though the underlying TS service should be the same for both vue-tsc and volar. Are you using the latest vue-tsc in package.json? `npm list vue-tsc`. It could be your vscode has updated volar automatically to the latest version (works) but your NPM vue-tsc stayed old or something.

@sethidden When i install the latest version of `vscode volar`, it not suggestion the `scoped slots props` type event I import `.vue` file directly in then same project.....After i fallback to `0.33.9`, same project is ok.

**Comment by Thy3634** [comment]
When I use `vue-tsc` generate dts from sfc (js + props comment)

```vue
<script>

export default {
    props: {
        /** comment */
        aProp: String
    }
}
</script>
```

output:
```ts
const Comp: DefineComponent<{any, ...... , {
    [x: string]: any;
}>
```

And  if remove comment or change this into `<script lang="ts">` ,it's ok.

**Comment by SavvyShah** [comment]
I'm using JSX in a Vue3 project. All the files are javascript `.jsx`. I assume we can't emit `.d.ts` file using `vue-tsc` for `.jsx` files. Is there a way I can create a `.d.ts` file manually? Is it safe to create it manually?

```jsx
// SelectedFilters.jsx
import VueTypes from 'vue-types';

const SelectedFilters = {
	name: 'SelectedFilters',
	props: {
		className: VueTypes.string.def(''),
		clearAllLabel: VueTypes.string.def('Clear All'),
		showClearAll: VueTypes.bool.def(true),
		resetToDefault: VueTypes.bool.def(false),
		clearAllBlacklistComponents: VueTypes.array,
		resetToValues: VueTypes.object,
	},
	render() {
                 // rendering logic using JSX
	},
};
```

**Comment by eric-gitta-moore** [comment]
For me, I don't really want to take over the rollup, swc or oxc at the lower level. I don't want to call the load and transform functions myself. Instead, I found that using chunkSplit can achieve the separation of each file.

ref: https://github.com/epicmaxco/vuestic-ui/blob/16459e2a5b67fd4cc2d795c438c4e4130710f4ee/packages/ui/build/common-config.ts#L98
